src/DotNetty.Handlers/Tls/SniHandler.cs

```diff
@@ -0,0 +1,340 @@
+
+                                                var idn = new IdnMapping()
+                                                {
+                                                    AllowUnassigned = true
+                                                };
+
+                                                hostname = idn.GetAscii(hostname);
+#if NETSTANDARD1_3
+                                                // TODO: netcore does not have culture sensitive tolower()
+                                                hostname = hostname.ToLowerInvariant();
+#else
+                                                hostname = hostname.ToLower(new CultureInfo("en-US"));
+#endif
+                                                this.Select(context, hostname);
+                                                return;
+                                            }
+                                            else
+                                            {
+                                                // invalid enum value
+                                                continueLoop = false;
+                                                break;
+                                            }
+                                        }
+
+                                        offset += extensionLength;
+                                    }
+                                }
+
+                                break;
+                            // Fall-through
+                            default:
+                                //not tls, ssl or application data, do not try sni
+                                continueLoop = false;
+                                break;
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    error = e;
+
+                    // unexpected encoding, ignore sni and use default
+                    if (Logger.DebugEnabled)
+                    {
+                        Logger.Warn($"Unexpected client hello packet: {ByteBufferUtil.HexDump(input)}", e);
+                    }
+                }
+
+                if (this.serverTlsSniSettings.DefaultServerHostName != null)
+                {
+                    // Just select the default certifcate
+                    this.Select(context, this.serverTlsSniSettings.DefaultServerHostName); 
+                }
+                else
+                {
+                    this.handshakeFailed = true;
+                    var e = new InvalidOperationException("Unable to determine server certicate", error);
+                    TlsUtils.NotifyHandshakeFailure(context, e);
+                    throw e;
+                }
+            }
+        }
+
+        void Select(IChannelHandlerContext context, string hostName)
+        {
+            Contract.Requires(hostName != null);
+            var getCertTask = this.serverTlsSniSettings.ServerCertificateSelector(hostName);
```

Comment:
pls switch to async / await. Yes, it will be an "async void" method, but you'll protect the body with try-catch and it'll be ok. It will reduce unnecessary complication of code + make sure continuation is dispatched back to executor.